### PR TITLE
Add rule for detecting predicates that always evaluates true

### PR DESCRIPTION
**Pull request by udonne**

### CODE REVIEW

File: TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.SqlServer.TransactSql.ScriptDom;
+using TSQLLint.Lib.Rules.Interface;
+
+namespace TSQLLint.Lib.Rules
+{
+    public class PredicateIsAlwaysTrue : TSqlFragmentVisitor, ISqlRule
+    {
+        public string RULE_NAME => "predicate-is-always-true";
+
+        public string RULE_TEXT => "Found the same on the left and the right side of a Predicate";
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs]: Found the same what? value, expression, string?

File: TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        
+        public override void Visit(BooleanComparisonExpression node)
+        {
+            {
+                var first = node.FirstExpression as Literal;
+                var second = node.SecondExpression as Literal;
+
+                if (first != null && second != null)
+                {
+                    if (string.Equals(first.Value, second.Value))
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs]: Not necessary to have a 2nd nested if - make it all one line

Comment (geoffbaker) [inline TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs]: Not necessary to have 2nd line - make it all 1 line

File: TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    {
+                        ErrorCallback(RULE_NAME, RULE_TEXT, node.StartLine, node.StartColumn);
+                    }
+                }
+            }
+            
+            {
+                var first = node.FirstExpression as ColumnReferenceExpression;
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs]: Looks like block will happen even if FirstExpression and SecondExpression are both Literals. If this is not the intent then need to exit routine if first and second are not null literals around line 33.

File: TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs [hunk 4]
```diff
@@ -20,32 +20,26 @@ public PredicateIsAlwaysTrue(Action<string, string, int, int> errorCallback)
...
-                }
-            }
-            
+            if (first != null && second != null && string.Equals(first.Value, second.Value))
+	    {
+	        ErrorCallback(RULE_NAME, RULE_TEXT, node.StartLine, node.StartColumn);
+                return;
+	    }
+
+            var first = node.FirstExpression as ColumnReferenceExpression;
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/PredicateIsAlwaysTrue.cs]: Change still doesn't address the fact that if first != null and second != null you don't want to continue to then try to see if they are ColumnReferenceExpression.

Honestly, I'd break this into two smaller private methods. One for Literal and one for Column Reference Expression both called from Visit. I'd make the return value a boolean indicating whether or not it was able to be translated to the corresponding value. Then you could just use short circuit logic to not call the second method if it is the first one. If a third shows up that pattern would still hold.

File: TSQLLint.Lib/Rules/FunctionInPredicateTest.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Linq;
+using Microsoft.SqlServer.TransactSql.ScriptDom;
+using TSQLLint.Lib.Rules.Interface;
+
+namespace TSQLLint.Lib.Rules
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/FunctionInPredicateTest.cs]: Why is file name called a Test when this looks to be actual code?

File: TSQLLint.Lib/Rules/FunctionInPredicateTest.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+using Microsoft.SqlServer.TransactSql.ScriptDom;
+using TSQLLint.Lib.Rules.Interface;
+
+namespace TSQLLint.Lib.Rules
+{
+    public class FunctionInPredicate : TSqlFragmentVisitor, ISqlRule
+    {
+        public string RULE_NAME => "function-in-predicate";
+
+        public string RULE_TEXT => "Don’t use functions in the WHERE or JOIN clause, they reduce performance.(index cannot be used)";
```
Comment (geoffbaker) [inline TSQLLint.Lib/Rules/FunctionInPredicateTest.cs]: add space before ( and move period till after )

File: TSQLLint.Tests/UnitTests/Rules/function-in-predicate/Test.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace TSQLLint.Tests.UnitTests.Rules
+{
+    public class FunctionInPredicateTests
+    {
+        private static readonly object[] testCases =
+        {
+            new object[]
+            {
+                "function-in-predicate", "function-in-predicate-where",  typeof(FunctionInPredicate), new List<RuleViolation>()
```
Comment (geoffbaker) [inline TSQLLint.Tests/UnitTests/Rules/function-in-predicate/Test.cs]: I like the always true/false naming convention for the others. Without this its harder to know if this is suppose to pass or fail

**Comment by udonne** [comment]
See it as hint for two useful rules. They work for my usage even if they are not perfect. Feel free to fix the issues.

**Pull request closed without merging**